application/src/main/java/org/togetherjava/tjbot/commands/basic/RoleSelectCommand.java

```diff
@@ -0,0 +1,319 @@
+                        return SelectOption.of(role.getName(), role.getId())
+                            .withEmoji((Emoji.fromUnicode(roleIcon.getEmoji())));
+                    }
+                })
+                .toList());
+    }
+
+    /**
+     * Creates an embedded message to send with the selection menu
+     *
+     * <p>
+     * </p>
+     *
+     * @param title for the embedded message. nullable {@link String}
+     * @param description for the embedded message. nullable {@link String}
+     * @return the formatted embed {@link MessageEmbed}
+     */
+    private static @NotNull MessageEmbed makeEmbed(@Nullable final String title,
+            @Nullable final CharSequence description) {
+
+        String effectiveTitle = (null == title) ? "Select your roles:" : title;
+
+        return new EmbedBuilder().setTitle(effectiveTitle)
+            .setDescription(description)
+            .setColor(AMBIENT_COLOR)
+            .build();
+    }
+
+    @Override
+    public void onSelectionMenu(@NotNull final SelectionMenuEvent event,
+            @NotNull final List<String> args) {
+
+        Guild guild = Objects.requireNonNull(event.getGuild(), "The given guild cannot be null");
+        List<SelectOption> selectedOptions = Objects.requireNonNull(event.getSelectedOptions(),
+                "The given selectedOptions cannot be null");
+
+
+        List<Role> selectedRoles = selectedOptions.stream()
+            .map(SelectOption::getValue)
+            .map(guild::getRoleById)
+            .filter(Objects::nonNull)
+            .filter(role -> guild.getSelfMember().canInteract(role))
+            .toList();
+
+
+        if (event.getMessage().isEphemeral()) {
+            handleNewRoleBuilderSelection(event, selectedRoles);
+        } else {
+            handleRoleSelection(event, selectedRoles, guild);
+        }
+    }
+
+    /**
+     * Handles selection of a {@link SelectionMenuEvent}
+     *
+     * @param event the <b>unacknowledged</b> {@link SelectionMenuEvent}
+     * @param selectedRoles the {@link Role roles} selected
+     * @param guild the {@link Guild}
+     */
+    private static void handleRoleSelection(final @NotNull SelectionMenuEvent event,
+            final @NotNull Collection<Role> selectedRoles, final Guild guild) {
+        Collection<Role> rolesToAdd = new ArrayList<>(selectedRoles.size());
+        Collection<Role> rolesToRemove = new ArrayList<>(selectedRoles.size());
+
+        event.getInteraction().getComponent().getOptions().stream().map(selectedOption -> {
+            Role role = guild.getRoleById(selectedOption.getValue());
+
+            if (null == role) {
+                handleNullRole(selectedOption);
+            }
+
+            return role;
+        }).filter(Objects::nonNull).forEach(role -> {
+            if (selectedRoles.contains(role)) {
+                rolesToAdd.add(role);
+            } else {
+                rolesToRemove.add(role);
+            }
+        });
```

Comment:
A bit complex to read, maybe helper methods could help, in particular for the `map` call.